File: api/src/main/java/com/bindglam/mint/MintConfiguration.java
package com.bindglam.mint;

import com.bindglam.config.Configuration;
import com.bindglam.config.Field;
import com.bindglam.config.complex.EnumField;
import com.bindglam.mint.database.SQLDatabaseType;

import java.io.File;

public final class MintConfiguration extends Configuration {
    public final Field<String> language = createPrimitiveField("language", "english");

    public final Database database = new Database();
    public final class Database {
        public final SQL sql = new SQL();
        public final class SQL {
            public final Field<SQLDatabaseType> type = createExtendedComplexField(() -> new EnumField<>("database.sql.type", SQLDatabaseType.SQLITE, SQLDatabaseType.class));

            public final SQLite sqlite = new SQLite();
            public final class SQLite {
                public final Field<Boolean> autoCommit = createPrimitiveField("database.sql.SQLITE.auto-commit", true);
                public final Field<Integer> validTimeout = createPrimitiveField("database.sql.SQLITE.valid-timeout", 500);
            }

            public final MySQL mysql = new MySQL();
            public final class MySQL {
                public final Field<String> url = createPrimitiveField("database.sql.MYSQL.url", "localhost:3306");
                public final Field<String> database = createPrimitiveField("database.sql.MYSQL.database", "minecraft");
                public final Field<String> username = createPrimitiveField("database.sql.MYSQL.username", "root");
                public final Field<String> password = createPrimitiveField("database.sql.MYSQL.password", "1234");
                public final Field<Integer> maxPoolSize = createPrimitiveField("database.sql.MYSQL.max-pool-size", 10);
            }
        }

        public final Redis redis = new Redis();
        public final class Redis {
            public final Field<Boolean> enabled = createPrimitiveField("database.redis.enabled", false);
            public final Field<String> host = createPrimitiveField("database.redis.host", "localhost");
            public final Field<Integer> port = createPrimitiveField("database.redis.port", 6379);
            public final Field<String> password = createPrimitiveField("database.redis.password", "1234");
            public final Field<Integer> timeout = createPrimitiveField("database.redis.timeout", 1000);
        }
    }

    public final Economy economy = new Economy();
    public final class Economy {
        public final Currency currency = new Currency();
        public final class Currency {
            public final Field<String> defaultCurrency = createPrimitiveField("economy.currency.default-currency", "won");
        }
    }

    public MintConfiguration(File file) {
        super(file);
    }
}

File: core/src/main/kotlin/com/bindglam/mint/MintConfiguration.kt
package com.bindglam.mint

import com.bindglam.config.Configuration
import com.bindglam.config.complex.EnumField
import com.bindglam.mint.manager.DatabaseManagerImpl
import java.io.File
import java.util.function.Supplier

class MintConfiguration(file: File) : Configuration(file) {
    val language = createPrimitiveField<String>("language", "english")!!

    val database = Database()
    inner class Database {
        val sql = SQL()
        inner class SQL {
            val type = createExtendedComplexField(Supplier { EnumField("database.sql.type", DatabaseManagerImpl.SQLDatabaseType.SQLITE, DatabaseManagerImpl.SQLDatabaseType::class.java) })!!

            val autoCommit = createPrimitiveField("database.sql.SQLITE.auto-commit", true)!!
            val validTimeout = createPrimitiveField("database.sql.SQLITE.valid-timeout", 500)!!

            val mysql = MySQL()
            inner class MySQL {
                val url = createPrimitiveField("database.sql.MYSQL.url", "localhost:3306")!!
                val database = createPrimitiveField("database.sql.MYSQL.database", "minecraft")!!
                val username = createPrimitiveField("database.sql.MYSQL.username", "root")!!
                val password = createPrimitiveField("database.sql.MYSQL.password", "1234")!!
                val maxPoolSize = createPrimitiveField("database.sql.MYSQL.max-pool-size", 10)!!
            }
        }

        val redis = Redis()
        inner class Redis {
            val enabled = createPrimitiveField("database.redis.enabled", false)!!
            val host = createPrimitiveField("database.redis.host", "localhost")!!
            val port = createPrimitiveField("database.redis.port", 6379)!!
            val password = createPrimitiveField("database.redis.password", "1234")!!
            val timeout = createPrimitiveField("database.redis.timeout", 1000)!!
            val syncInterval = createPrimitiveField("database.redis.sync-interval", 60)!!
        }
    }

    val economy = Economy()
    inner class Economy {
        val currency = Currency()
        inner class Currency {
            val defaultCurrency = createPrimitiveField("economy.currency.default-currency", "won")!!
        }
    }
}